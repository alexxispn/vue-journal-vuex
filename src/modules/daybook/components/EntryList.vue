<template>
  <div class="entry-list-container">
    <div class="px-2 pt-2">
      <input type="text" class="form-control" placeholder="Search..." v-model="term"/>
    </div>
    <div class="entry-scrollarea">
      <div class="mt-2 d-flex flex-column">
        <button class="btn btn-primary mx-3" @click="toNewEntry">
          <i class="fa fa-plus circle"></i>
          New Entry
        </button>
      </div>
      <EntryItem v-for="entry in entriesByTerm" :key="entry.id" :entry="entry"/>
    </div>
  </div>
</template>

<script>
import {defineAsyncComponent} from "vue";
import {mapGetters} from "vuex";

export default {
  name: "EntryList",
  components: {
    EntryItem: defineAsyncComponent(() => import(/* webpackChunkName: "EntryItem" */ "./EntryItem.vue"))
  },
  computed: {
    ...mapGetters("journal", ['getEntriesByTerm']),
    entriesByTerm() {
      return this.getEntriesByTerm(this.term);
    }
  },
  methods: {
    toNewEntry() {
      this.$router.push({name: 'entry', params: {id: 'new'}});
    }
  },
  data() {
    return {
      term: ''
    }
  }
}

</script>

<style lang="scss" scoped>

.entry-list-container {
  border-right: 1px solid #2c3e50;
  height: calc(100vh - 56px);
}

.entry-scrollarea {
  height: calc(100vh - 56px - 50px);
  overflow-y: scroll;
}

</style>
