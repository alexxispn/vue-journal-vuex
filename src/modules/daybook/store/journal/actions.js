import journalApi from "@/api/journalApi";

export const loadEntries = async ({commit}) => {
    const {data} = await journalApi.get("/entries.json")
    if (!data) {
        commit("setEntries", []);
    } else {
        const entries = [];
        for (let id of Object.keys(data)) {
            entries.push({
                id,
                ...data[id]
            });
        }
        commit("setEntries", entries);
        return entries;
    }
}

export const updateEntry = async ({commit}, entry) => {
    const newEntry = {
        ...entry
    }
    delete newEntry.id;
    await journalApi.put(`/entries/${entry.id}.json`, newEntry);
    commit("updateEntry", newEntry);
    return entry.id;
}

export const addEntry = async ({commit}, entry) => {
    const {data} = await journalApi.post("/entries.json", entry);
    commit("addEntry", {...entry, id: data.name});
    return data.name;
}


export const deleteEntry = async ({commit}, id) => {
    await journalApi.delete(`/entries/${id}.json`)
    commit("deleteEntry", id);
    return null
}
