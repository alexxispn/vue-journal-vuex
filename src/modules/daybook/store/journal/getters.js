export const getEntriesByTerm = (state) => (term) => {
    return state.entries.filter((entry) => {
        return entry.text.toLowerCase().includes(term.toLowerCase()) || entry.date.includes(term)
    })
}

export const getEntryById = (state) => (id) => {
    const entry = state.entries.find((entry) => entry.id === id)
    if (entry) {
        return entry
    }
    return null
}
