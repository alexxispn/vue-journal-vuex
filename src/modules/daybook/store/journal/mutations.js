export const setEntries = (state, entries) => {
    state.entries = entries;
}

export const updateEntry = (state, entry) => {
    state.entries = state.entries.map(e => e.id === entry.id ? entry : e);
}
