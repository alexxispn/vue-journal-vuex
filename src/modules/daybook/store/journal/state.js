export default () => ({
    isLoading: true,
    entries: [
        {
            id: '1',
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quod.',
            picture: null
        },
        {
            id: '2',
            date: new Date().toDateString(),
            text: 'Ipsum dolor sit amet consectetur adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quod.',
            picture: null
        },
        {
            id: '3',
            date: new Date().toDateString(),
            text: 'Dolor Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam, quod.',
            picture: null
        }
    ]
})
