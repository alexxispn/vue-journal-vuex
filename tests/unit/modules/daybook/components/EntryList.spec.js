import {shallowMount} from "@vue/test-utils";
import {createStore} from "vuex";
import journal from "@/modules/daybook/store/journal";
import EntryList from "@/modules/daybook/components/EntryList";
import {journalState} from "../../../mock-data/journal-state";

const createVuexStore = (initialState) =>
    createStore({
        modules: {
            journal: {
                ...journal,
                state: {...initialState}
            }
        }
    });

describe("EntryList component", () => {
    const store = createVuexStore(journalState);
    const mockRouter = {
        push: jest.fn()
    }
    const wrapper = shallowMount(EntryList, {
        global: {
            mocks: {
                $router: mockRouter
            },
            plugins: [store]
        }
    });
    it("should match snapshot", () => {
        expect(wrapper.html()).toMatchSnapshot();
    });
    it("should call getEntriesByTerm without term and show 2 entries", () => {
        expect(wrapper.findAll("entry-item-stub").length).toBe(2);
    });
    it("should call getEntriesByTerm with term and show 1 entry", async () => {
        const input = wrapper.find("input");
        await input.setValue("Hola");
        expect(wrapper.findAll("entry-item-stub").length).toBe(1);
    });
    it("should redirect to new entry when click on new entry button", async () => {
        const button = wrapper.find("button");
        await button.trigger("click");
        expect(mockRouter.push).toHaveBeenCalledWith({
            name: "entry",
            params: {
                id: "new"
            }
        });
    });
});
