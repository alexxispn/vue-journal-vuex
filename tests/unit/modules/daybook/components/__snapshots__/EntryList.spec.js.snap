// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryList component should match snapshot 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Search..."></div>
  <div class="entry-scrollarea">
    <div class="mt-2 d-flex flex-column"><button class="btn btn-primary mx-3"><i class="fa fa-plus circle"></i> New Entry </button></div>
    <entry-item-stub entry="[object Object]"></entry-item-stub>
    <entry-item-stub entry="[object Object]"></entry-item-stub>
  </div>
</div>
`;
